File: Backend/src/main/java/com/sencarmarket/module/vehicule/mapper/VehiculeMapper.java
package com.sencarmarket.module.vehicule.mapper;

import com.sencarmarket.module.vehicule.dto.VehiculeResponse;
import com.sencarmarket.module.vehicule.entity.Vehicule;
import com.sencarmarket.module.vehicule.repository.PhotoVehiculeRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Mapper pour convertir les entités Vehicule en DTOs
 * Respecte le principe SRP en centralisant la logique de mapping
 */
@Component
@RequiredArgsConstructor
public class VehiculeMapper implements IVehiculeMapper {

    private final PhotoVehiculeRepository photoVehiculeRepository;

    /**
     * Convertit une entité Vehicule en VehiculeResponse
     */
    public VehiculeResponse toResponse(Vehicule vehicule) {
        if (vehicule == null) {
            return null;
        }

        List<String> photos = photoVehiculeRepository.findByVehiculeId(vehicule.getId()).stream()
                .map(photo -> photo.getUrl())
                .collect(Collectors.toList());

        return VehiculeResponse.builder()
                .id(vehicule.getId())
                .marque(vehicule.getMarque() != null ? vehicule.getMarque().getNom() : null)
                .modele(vehicule.getModele() != null ? vehicule.getModele().getNom() : null)
                .anneeFabrication(vehicule.getAnneeFabrication())
                .kilometrage(vehicule.getKilometrage())
                .carburant(vehicule.getCarburant() != null ? vehicule.getCarburant().getNom() : null)
                .boiteVitesse(vehicule.getBoiteVitesse() != null ? vehicule.getBoiteVitesse().getNom() : null)
                .couleur(vehicule.getCouleur())
                .prixVente(vehicule.getPrixVente())
                .description(vehicule.getDescription())
                .numeroVin(vehicule.getNumeroVin())
                .immatriculation(vehicule.getImmatriculation())
                .statut(vehicule.getStatut() != null ? vehicule.getStatut().name() : null)
                .prixNegociable(vehicule.getPrixNegociable())
                .certifie(vehicule.getCertifie())
                .photosUrls(photos)
                .estBoost(vehicule.getEstBoost())
                .boostDebut(vehicule.getBoostDebut())
                .boostFin(vehicule.getBoostFin())
                .vues(vehicule.getVues())
                .createdAt(vehicule.getCreatedAt())
                .build();
    }

    /**
     * Convertit une liste d'entités en liste de DTOs
     */
    public List<VehiculeResponse> toResponseList(List<Vehicule> vehicules) {
        if (vehicules == null) {
            return null;
        }
        return vehicules.stream()
                .map(this::toResponse)
                .collect(Collectors.toList());
    }
}

File: Backend/src/modules/vehicule/services/vehicule.mapper.ts
import { Injectable } from '@nestjs/common';

import { VehiculeResponseDto } from '../dto/vehicule-response.dto';
import { VehiculeRecord } from '../vehicule.models';

@Injectable()
export class VehiculeMapper {
  toVehiculeResponse(vehicule: VehiculeRecord, estFavori: boolean): VehiculeResponseDto {
    const marque = (vehicule.marque as { nom?: string | null } | null)?.nom ?? null;
    const marqueId = vehicule.marque_id;
    const modele = (vehicule.modele as { nom?: string | null } | null)?.nom ?? null;
    const modeleId = vehicule.modele_id;
    const carburant = (vehicule.carburant as { nom?: string | null } | null)?.nom ?? null;
    const carburantId = vehicule.carburant_id;
    const boiteVitesse = (vehicule.boite_vitesse as { nom?: string | null } | null)?.nom ?? null;
    const boiteVitesseId = vehicule.boite_vitesse_id;
    const proprietaire = vehicule.utilisateur;
    const photos = ((vehicule.photo_vehicule as Array<{ url: string }> | undefined) ?? []).map(
      (p) => p.url,
    );

    return {
      id: vehicule.id,
      marque,
      marqueId,
      modele,
      modeleId,
      anneeFabrication: vehicule.annee_fabrication,
      kilometrage: vehicule.kilometrage,
      carburant,
      carburantId,
      boiteVitesse,
      boiteVitesseId,
      couleur: vehicule.couleur,
      prixVente: vehicule.prix_vente ? String(vehicule.prix_vente) : null,
      description: vehicule.description,
      numeroVin: vehicule.numero_vin,
      immatriculation: vehicule.immatriculation,
      statut: vehicule.statut,
      prixNegociable: vehicule.prix_negociable,
      certifie: vehicule.certifie,
      titre: vehicule.titre,
      nombrePortes: vehicule.nombre_portes,
      nombrePlaces: vehicule.nombre_places,
      cylindree: vehicule.cylindree,
      puissanceFiscale: vehicule.puissance_fiscale,
      estGarantie: vehicule.est_garantie,
      garantieMois: vehicule.garantie_mois,
      photosUrls: photos,
      estBoost: vehicule.est_boost,
      boostDebut: vehicule.boost_debut,
      boostFin: vehicule.boost_fin,
      vues: vehicule.vues,
      nombreFavoris: vehicule.nombre_favoris,
      estFavori,
      proprietaireNom: proprietaire?.nom ?? null,
      proprietaireId: vehicule.proprietaire_id,
      proprietaireTelephone: proprietaire?.telephone ?? null,
      proprietaireEmail: proprietaire?.email ?? null,
      createdAt: vehicule.created_at,
    };
  }
}
